### PR TITLE
Library improvements

### PR DESCRIPTION
**Pull request by neurogoo**
Add way to add physical books to library and clean up code

### CODE REVIEW

File: library-app/src/Futurice/App/Library/Pages/EditItemPage.hs [hunk 1]
```diff
@@ -36,7 +36,24 @@ editItemPage (Left info) = page_ "Edit book information" (Nothing :: Maybe Nav)
                 th_ "Amazon link"
                 td_ $ input_ [ name_ "amazon-link", type_ "text", value_ (info ^. bookAmazonLink)]
         button_ [ class_ "button success", data_ "futu-action" "submit" ] $ "Edit"
-editItemPage (Right info) =  page_ "Edit board game information" (Nothing :: Maybe Nav) $ do
+    h3_ "Books"
+    ul_ [class_ ""] $ do
+        for_ (sortBy (\(_,_,lib1) (_,_,lib2) -> librarySelectSortOrder lib1 lib2) books) $ \(itemid, status, library) ->
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Pages/EditItemPage.hs]:
```haskell
librarySelectSortOrder `on` view _3
```

File: library-app/src/Futurice/App/Library/Types/BorrowRequest.hs [hunk 2]
```diff
@@ -20,6 +22,10 @@ data BorrowRequest = BorrowRequest
 
 deriveGeneric ''BorrowRequest
 
-deriveVia [t| FromJSON BorrowRequest `Via` Sopica BorrowRequest |]
+instance FromJSON BorrowRequest where
+    parseJSON (Object v) =  BorrowRequest
+        <$> v .: "book"
+        <*> (libraryFromText <$> v .: "library")
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Types/BorrowRequest.hs]: can't you add `FromJSON Library` and use just `v .: "library"`

Comment (phadej) [reply library-app/src/Futurice/App/Library/Types/BorrowRequest.hs]: and then `deriveVia` might work?

File: library-app/src/Futurice/App/Library/Types/AddItemRequest.hs [hunk 3]
```diff
@@ -45,13 +49,22 @@ data AddBoardGameInformation = AddBoardGameInformation
...
+librarySelectSortOrder :: Library -> Library -> Ordering
+librarySelectSortOrder a b = case (a,b) of
                         (Elibrary, Elibrary) -> EQ
                         (Elibrary, _) -> GT
                         (_, Elibrary) -> LT
                         _ -> compare (libraryToText a) (libraryToText b)
-                    usedLibs = (filter (/= OfficeLibrary offOther) . filter (/= UnknownLibrary)) allLibraries
+
+usedLibraries :: [Library]
+usedLibraries = let usedLibs = (filter (/= OfficeLibrary offOther) . filter (/= UnknownLibrary)) allLibraries
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: make a predicate `is???Library :: Library -> Bool`

File: library-app/src/Futurice/App/Library/Types/AddItemRequest.hs [hunk 4]
```diff
@@ -45,13 +49,22 @@ data AddBoardGameInformation = AddBoardGameInformation
...
-usedLibraries :: [Library]
-usedLibraries = let librarySelectSortOrder a b = case (a,b) of
+data AddItemRequest = AddItemRequest
+    { _addItemLibrary                :: !Library
+    , _addItemBookInformationId      :: !(Maybe BookInformationId)
+    , _addItemBoardGameInformationId :: !(Maybe BoardGameInformationId)
+    }
+    deriving (Show, Generic, ToSchema)
+
+librarySelectSortOrder :: Library -> Library -> Ordering
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: add a comment: alphabetical, but elibrary last

File: library-app/src/Futurice/App/Library/Types/AddItemRequest.hs [hunk 5]
```diff
@@ -45,13 +49,22 @@ data AddBoardGameInformation = AddBoardGameInformation
...
     deriving Show
 
-usedLibraries :: [Library]
-usedLibraries = let librarySelectSortOrder a b = case (a,b) of
+data AddItemRequest = AddItemRequest
+    { _addItemLibrary                :: !Library
+    , _addItemBookInformationId      :: !(Maybe BookInformationId)
+    , _addItemBoardGameInformationId :: !(Maybe BoardGameInformationId)
+    }
+    deriving (Show, Generic, ToSchema)
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: you don't want default `ToSchema` (via GHC.Generics)

Comment (phadej) [reply library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: use explict `deriving stock` ...

File: library-app/src/Futurice/App/Library/Types/BorrowRequest.hs (hunk 2 above)
Comment (neurogoo) [reply library-app/src/Futurice/App/Library/Types/BorrowRequest.hs]: Ah didn't realize the problem was this

File: library-app/src/Futurice/App/Library/Types/AddItemRequest.hs (hunk 5 above)
Comment (neurogoo) [reply library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: Umm, but ToSchema is not stoack derivable class?

Comment (phadej) [reply library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: It isn't. But you don't want derive it via `DeriveAnyClass`

Comment (neurogoo) [reply library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: but it isn't allowed to derive via stock so I should then just handwrite it?

Comment (phadej) [reply library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: Hmm. Why you even need the `ToSchema`.  Does multipart requires it?

If so, than you could write `deriving anyclass (ToSwagger) -- we don't really care about the schema, just that there's one`

Comment (neurogoo) [reply library-app/src/Futurice/App/Library/Types/AddItemRequest.hs]: Hmm I am pretty sure something needed it, but seems that it is not used anymore ^^'

**Pull request merged**